Add UI code for configuring subgraphNode widgets

AustinMroz opened the pull request:
The third PR for managing display of widgets on subgraph nodes. This is the one that actually makes the functionality usable and user visible.

Adds
- A right-side modal for configuring which widgets are promoted, accessed by right click or selection toolbar
  - This menu allows for re-arranging widget order by dragging and dropping.
- Indicators inside the subgraph for which widgets have been promoted.
- Context menu options for promoting or demoting widget inside of a subgraph.
<img width="767" height="694" alt="image" src="https://github.com/user-attachments/assets/4f78645d-7b26-48ba-8c49-78f4807e89e8" />
<img width="784" height="435" alt="image" src="https://github.com/user-attachments/assets/7005c730-a732-481e-befb-57019a8a31a7" />


Known issues
- Some preview widgets are not added to a node until a draw operation occurs. The code does not yet have a way of determining which nodes should have draw operations forced to facilitate initial widget creation.

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-5826-Add-UI-code-for-configuring-subgraphNode-widgets-27c6d73d36508146accbf395e5bcd36a) by [Unito](https://www.unito.io)

AustinMroz committed:
Add UI code for configuring subgraphNode widgets

christian-byrne reviewed (commented):
This looks good, I left a lot of very minor comments on the first pass through. Have not tested yet!

AustinMroz committed:
Better property reactivity, nits

triggerUpdate has been removed and access of the proxyWidgets property
has been made reactive. This handles triggering nicely, but note that
the getter here is not cached. This doesn't really matter since the
accessed property is also cached.

Also remove export of the CANVAS_IMAGE_PREVIEW_WIDGET type. Different
classes of previews have different names, and needing to manually import
each of the types makes the code fragile to future updates. instead,
these dynamically added widgets are identified by their leading double $

AustinMroz committed:
Move to unstyled+tailwind

The vars which require fallbacks are a little clunky, but should be
safer for future revision than raw hexcodes

AustinMroz committed:
Localization fixes

AustinMroz committed:
Use emit instead of callback

AustinMroz committed:
Use utility functions and emit

AustinMroz commented on src/services/litegraphService.ts:
```diff
@@ -824,18 +826,88 @@ export const useLitegraphService = () => {
-            this.graph.unpackSubgraph(this)
+        options.unshift(
+          {
+            content: 'Edit Subgraph Widgets',
+            callback: () => {
+              showSubgraphNodeDialog()
+            }
+          },
+          {
+            content: 'Unpack Subgraph',
```
Done, but a bit of a rabbit hole. Translation is already being applied to context menu option fields, so adding translation here simply requires adding the field to locales.

As part of this change, I've removed the 🆕 tag from the "Convert to Subgraph" option. It was preventing translation.

AustinMroz commented on src/core/graph/subgraph/SubgraphNode.vue:
1. This was originally a goal. Took more work than expected, but the result with emit change is clean.
2. Nope. :+1:

AustinMroz commented on src/core/graph/subgraph/SubgraphNode.vue:
```diff
@@ -0,0 +1,288 @@
+  </div>
+  <div v-if="recommendedWidgets.length" class="justify-center flex py-4">
+    <Button size="small" @click.stop="showRecommended">
+      {{ t('subgraphStore.showRecommended') }}
```
There was a little bit of back and forth with design over whether the button should be kept. There's currently a slight lean towards "yes".

In the future, I think I'll commit to marking items which should be translated with a placeholder tag like `t('[PH]Show Recommended Widgets')`. Not wanting to commit to locale changes early is fine. Having mixed, miss-able placeholders is problematic.

AustinMroz commented:
Current playwritght test failures are from removing the 🆕 character from "Convert to Subgraph" to fix translation and are expected.

DrJKL commented on src/core/graph/subgraph/SubgraphNodeWidget.vue:
```diff
@@ -0,0 +1,36 @@
+</script>
+<template>
+  <div
+    class="flex pt-1 pr-4 pb-1 pl-0 break-all rounded bg-[var(--p-dialog-background,#202020)] items-center gap-1"
```
Can we use one of our theme colors for the background here?

AustinMroz replied:
Sure.

It doesn't inspire confidence that for the closest match (`charcoal-800`) the codebase is split across 4 different values for the now required light theme equivalent (`gray-50, gray-100, white, pure-white`), but I'd rather not be the one to fry that fish.

DrJKL commented on src/core/graph/subgraph/SubgraphNodeWidget.vue:
```diff
@@ -0,0 +1,36 @@
+</script>
+<template>
+  <div
+    class="flex pt-1 pr-4 pb-1 pl-0 break-all rounded bg-[var(--p-dialog-background,#202020)] items-center gap-1"
```
optional:
```suggestion
    class="flex py-1 pr-4 pl-0 break-all rounded bg-[var(--p-dialog-background,#202020)] items-center gap-1"
```

DrJKL commented on src/core/graph/subgraph/SubgraphNodeWidget.vue:
```diff
@@ -0,0 +1,36 @@
+  <div
+    class="flex pt-1 pr-4 pb-1 pl-0 break-all rounded bg-[var(--p-dialog-background,#202020)] items-center gap-1"
+  >
+    <div class="size-4">
+      <i-lucide:grip-vertical v-if="isDraggable" />
+    </div>
```
Do you need the wrapper div in this case?

AustinMroz replied:
With the power of class form, it is not.

DrJKL commented on src/lib/litegraph/src/types/widgets.ts:
```diff
@@ -306,6 +306,12 @@ export interface IBaseWidget<
 
   hidden?: boolean
   advanced?: boolean
+  /**
+   * Set if the node is displayed on the parent subgraphNode
+   * Promoted widgets have a green border
+   * @readonly [Computed] This property is computed on graph change
+   */
```
This is a lot of comment...
What happens if we change the theme color for the border here?

DrJKL commented on src/scripts/domWidget.ts:
```diff
@@ -160,16 +160,24 @@ abstract class BaseDOMWidgetImpl<V extends object | string>
       const originalFillStyle = ctx.fillStyle
       ctx.beginPath()
       ctx.fillStyle = LiteGraph.WIDGET_BGCOLOR
+      let margin = this.margin
```
nit: I have a general preference against mutable state like this.

AustinMroz commented on src/components/graph/selectionToolbox/ConfigureSubgraph.vue:
```diff
@@ -0,0 +1,21 @@
+    @click="showSubgraphNodeDialog"
+  >
+    <template #icon>
+      <i-lucide:settings2 />
```
Pretty sure it wasn't merged when I first started the feature branch. Thanks for the reminder. I prefer it too!

AustinMroz commented on src/core/graph/subgraph/proxyWidget.ts:
```diff
@@ -89,16 +113,18 @@ function addProxyWidget(
-    label: name,
+    afterQueued: undefined,
+    computedHeight: undefined,
     isProxyWidget: true,
-    y: 0,
+    label: name,
     last_y: undefined,
-    width: undefined,
-    computedHeight: undefined,
-    afterQueued: undefined,
+    name,
+    node: subgraphNode,
     onRemove: undefined,
-    node: subgraphNode
+    promoted: undefined,
+    serialize: false,
+    width: undefined,
+    y: 0
```
It puts the Overlay properties first , followed by the IBaseWidget properties in alphabetical order.

I'll add a line gap between the two and make the first group alphabetical

AustinMroz committed:
Fix nits

AustinMroz committed:
Remove string serialization of proxyWidgets prop

AustinMroz committed:
Update comment to not reference color

The actual color used has been changed so the comment was misleading

AustinMroz committed:
Further icon cleanup

AustinMroz committed:
Use utility functions and more descriptive names

AustinMroz committed:
swap color variables

AustinMroz committed:
Deduplicate recommended, test typing, ordering

Code for filtering recommended exists as a single function in
proxyWidgetUtils.

The "as unknown" cast in tests has been replaced with an "as partial"
cast which provides some actual type safety

refs, computeds, and functions have been sorted

AustinMroz committed:
Use ternary for widgetmargin

AustinMroz committed:
Remove vuedraggable dependency

The library swapped to clearly isn't designed with vue in mind. I'm
still verifying that it's not creating problems with element lifetime,
but in testing, it seems to properly handle all edge cases

AustinMroz committed:
Condense display, swap to watchDebounced

AustinMroz committed:
Fix breakage of vue caching with draggableList

AustinMroz committed:
Fix broken import

AustinMroz committed:
Re-Add watchDebounced

The "Show Recommended" button would sometimes prevent dragging.

AustinMroz commented on src/core/graph/subgraph/SubgraphNode.vue:
```diff
@@ -0,0 +1,284 @@
+      >
+        <SubgraphNodeWidget
+          :node-title="widgetItem[0].title"
+          :widget-name="widgetItem[1].name"
```
Requires some repacking, but it's better.

AustinMroz committed:
Update dialog styling

AustinMroz commented on src/lib/litegraph/src/infrastructure/LGraphCanvasEventMap.ts:
```diff
@@ -14,6 +15,11 @@ export interface LGraphCanvasEventMap {
     /** The old active graph, or `null` if there was no active graph. */
     oldGraph: LGraph | Subgraph | null | undefined
   }
+  'subgraph-opened': {
+    subgraph: Subgraph
+    closingGraph: LGraph
+    fromNode: SubgraphNode
+  }
```
Agreed. I've been starting to think real hard about "How a user would expect linked subgraphs to work" and "How this could actually be accomplished in code.".

Here, the subgraph navigation stack is insufficient because it just "searches" for a route to the current subgraph every time the graph changes. Things would be much nicer if it properly tracked parent subgraphNodes and could be used as the single source of truth.

AustinMroz commented on src/core/graph/subgraph/proxyWidgetUtils.ts:
```diff
@@ -0,0 +1,126 @@
+  node: LGraphNode,
+  parents: SubgraphNode[]
+) {
+  for (const parent of parents) pushWidgets(parent, [`${node.id}`, widget.name])
```
I keep returning to this one, but I neither see a way it could be cleaned up, or how the number of mutations is anything short of minimal
- The code *really should* account for linked subgraphs even if the length of parent nodes is expected to always be 1
- Each parent has the singular mutation explicitly described by the function name applied.
- forEach is verboten

DrJKL replied:
- You get each parent
- You call pushWidgets
- pushWidgets calls getProxyWidgets
- getProxyWidgets calls parseProxyWidgets
- parseProxyWidgets goes into string land to try to deserialize any existing widgets (using safeParse, but also throwing its own error sometimes)
- Now you have, presumably, an array of _something_ (probably string,string tuples?)
- You dump new values in there (where I'd probably expect a uniqueness constraint, but I don't know if that's valid)
- You then stringify again


`demoteWidgets` has a similar structure but looks different and does things inline instead of abstracting to a function

It's just kind of a lot to have to trace for the logic here.
I still don't understand why we're ever stringifying outside of a serialization context.

DrJKL replied:
The functional form would be more like:
Given an existing set of "proxy widget" tuples and a target, return a new set that either adds or removes the target widget.
Then you could use that pure function across the relevant nodes.
It also makes undo/redo a lot more straightforward if you can stack these sorts of operations.

AustinMroz committed:
Manually update browser test expectations

AustinMroz committed:
Improve documentation for promoted widget property

AustinMroz committed:
Update spacing for topbar setting

When the "Opened workflows position" setting is set the default
"Topbar", additional spacing is required to ensure the menu doesn't
overlap with the run button.

AustinMroz committed:
Fix DOMWidget clipping and border under LOD

AustinMroz committed:
Manually fix other broken test screenshot

AustinMroz committed:
Marginally improved organization

AustinMroz committed:
Auto update previews, computedDisabled, autodemote

Previews on linked nodes and computedDisabled status is automatically
updated on state change.

A promoted Widget which is computed as disabled is automatically removed
from the node

AustinMroz committed:
Fix third playwright text

AustinMroz commented:
I have no blocking objections. There's a couple of tougher functional issues raised by Pablo I'm still iterating on, but everything here has been addressed.

The new browser test expectations workflow failed, so I've just been manually updating from test failures. This third commit should be sufficient to get tests passing, otherwise we may need to toggle the label to force the workflow to re-run
